src/app/encoder.cpp

```diff
@@ -2467,6 +2479,259 @@ PacketBufferHandle encodeLevelControlClusterReadClusterRevisionAttribute(Endpoin
     COMMAND_FOOTER();
 }
 
+/*----------------------------------------------------------------------------*\
+| Cluster NetworkProvisioning                                         | 0xAAAA |
+|------------------------------------------------------------------------------|
+| Commands:                                                           |        |
+| * AddThreadNetwork                                                  |   0x06 |
+| * AddWiFiNetwork                                                    |   0x02 |
+| * DisableNetwork                                                    |   0x0E |
+| * EnableNetwork                                                     |   0x0C |
+| * GetLastNetworkProvisioningResult                                  |   0x10 |
+| * RemoveNetwork                                                     |   0x0A |
+| * ScanNetworks                                                      |   0x00 |
+| * UpdateThreadNetwork                                               |   0x08 |
+| * UpdateWiFiNetwork                                                 |   0x04 |
+|------------------------------------------------------------------------------|
+| Attributes:                                                         |        |
+| * ClusterRevision                                                   | 0xFFFD |
+\*----------------------------------------------------------------------------*/
+
+/*
+ * Command AddThreadNetwork
+ */
+PacketBufferHandle encodeNetworkProvisioningClusterAddThreadNetworkCommand(EndpointId destinationEndpoint,
+                                                                           uint8_t * operationalDataset, uint64_t breadcrumb,
+                                                                           uint32_t timeoutMs)
+{
+    COMMAND_HEADER("AddThreadNetwork", NWPROV_CLUSTER_ID);
+    size_t operationalDatasetStrLen = strlen(reinterpret_cast<char *>(operationalDataset));
+    if (!CanCastTo<uint8_t>(operationalDatasetStrLen))
+    {
+        ChipLogError(Zcl, "Error encoding %s command. String too long: %d", kName, operationalDatasetStrLen);
+        return PacketBufferHandle();
+    }
+    buf.Put8(kFrameControlClusterSpecificCommand)
+        .Put8(kSeqNum)
+        .Put8(ZCL_ADD_THREAD_NETWORK_COMMAND_ID)
+        .Put(static_cast<uint8_t>(operationalDatasetStrLen))
+        .Put(reinterpret_cast<char *>(operationalDataset))
```

Comment:
Just a question.. Why we send Operational Dataset in a form of a string and not in the binary form of Operational Dataset TLV?

Reply:
@vivien-apple I think we need BYTE_STRING type support in encoder and dispatcher. (item 1 in #4503 )